File: ios/PacerKit/Sources/PacerKit/Sequencer.swift
import Foundation

/// Playlist intelligence: given a target cadence and a track pool, rank
/// tracks by how well they'd serve the runner *right now* and build an
/// adaptive "Up Next" queue.
public struct SequencerWeights: Sendable {
    /// Weight of the BPM compatibility score.
    public var bpmMatch: Double = 1
    /// Max penalty for a track played recently (scaled by recency).
    public var recentlyPlayed: Double = 0.45
    /// Penalty for a track the user skipped this session.
    public var skipped: Double = 0.75
    /// Max penalty for a jarring perceived-BPM jump from the current track.
    public var transition: Double = 0.2
    /// Perceived-BPM jump (SPM) at which the transition penalty saturates.
    public var transitionSaturationSpm: Double = 30
    /// Flat score for tracks with unknown BPM (kept, but ranked low).
    public var unknownBpmScore: Double = 0.15

    public init() {}
}

public struct SequencerContext: Sendable {
    public var targetSpm: Double
    /// Track currently playing (excluded from ranking, used for transition cost).
    public var currentTrack: PacerTrack?
    /// Track ids played recently, most recent last.
    public var recentlyPlayedIDs: [String]
    /// Track ids the user skipped this session.
    public var skippedIDs: [String]
    public var bpmPreferences: BpmMatchPreferences
    public var weights: SequencerWeights

    public init(
        targetSpm: Double,
        currentTrack: PacerTrack? = nil,
        recentlyPlayedIDs: [String] = [],
        skippedIDs: [String] = [],
        bpmPreferences: BpmMatchPreferences = BpmMatchPreferences(),
        weights: SequencerWeights = SequencerWeights()
    ) {
        self.targetSpm = targetSpm
        self.currentTrack = currentTrack
        self.recentlyPlayedIDs = recentlyPlayedIDs
        self.skippedIDs = skippedIDs
        self.bpmPreferences = bpmPreferences
        self.weights = weights
    }
}

public struct RankedTrack: Sendable, Identifiable {
    public let track: PacerTrack
    public let score: Double
    public let match: BpmMatch?
    public let reasons: [String]

    public var id: String { track.id }
}

public enum Sequencer {
    /// Rank every candidate track for the given context, best first.
    public static func rankTracks(_ tracks: [PacerTrack], context: SequencerContext) -> [RankedTrack] {
        let weights = context.weights
        let recent = context.recentlyPlayedIDs
        let skipped = Set(context.skippedIDs)
        let currentMatch = context.currentTrack.flatMap {
            BpmMatcher.match(cadenceSpm: context.targetSpm, trackBpm: $0.bpm, preferences: context.bpmPreferences)
        }

        var ranked: [RankedTrack] = []
        for track in tracks {
            if let current = context.currentTrack, track.id == current.id { continue }

            var reasons: [String] = []
            let match = BpmMatcher.match(
                cadenceSpm: context.targetSpm, trackBpm: track.bpm, preferences: context.bpmPreferences
            )
            var score: Double

            if let match {
                score = match.score * weights.bpmMatch
                reasons.append(
                    String(
                        format: "%.0f perceived BPM (%gx) vs %.0f SPM target",
                        match.perceivedBpm, match.multiplier, context.targetSpm
                    )
                )

                // Avoid whiplash transitions between consecutive tracks.
                if let currentMatch {
                    let jump = abs(match.perceivedBpm - currentMatch.perceivedBpm)
                    let penalty = min(1, jump / weights.transitionSaturationSpm) * weights.transition
                    if penalty > 0.01 {
                        score -= penalty
                        reasons.append(String(format: "transition jump %.0f BPM from current track", jump))
                    }
                }
            } else {
                score = weights.unknownBpmScore
                reasons.append("BPM unknown")
            }

            if let recentIndex = recent.lastIndex(of: track.id) {
                // Most recently played ⇒ full penalty; decays for older plays.
                let recency = Double(recentIndex + 1) / Double(recent.count)
                score -= weights.recentlyPlayed * recency
                reasons.append("played recently")
            }

            if skipped.contains(track.id) {
                score -= weights.skipped
                reasons.append("skipped this session")
            }

            ranked.append(RankedTrack(track: track, score: score, match: match, reasons: reasons))
        }

        return ranked.sorted { $0.score > $1.score }
    }

    /// Build the visible "Up Next" queue (top N ranked tracks).
    public static func buildQueue(
        _ tracks: [PacerTrack], context: SequencerContext, length: Int = 5
    ) -> [RankedTrack] {
        Array(rankTracks(tracks, context: context).prefix(length))
    }

    /// Pick the next track to play (or nil if nothing is available).
    public static func pickNextTrack(_ tracks: [PacerTrack], context: SequencerContext) -> RankedTrack? {
        rankTracks(tracks, context: context).first
    }
}

File: src/core/sequencer/sequencer.ts
import type { Track } from "../shared/types";
import { getBpmMatch, type BpmMatch, type BpmMatchPreferences } from "../bpm/match";

/**
 * Playlist intelligence: given a target cadence and a playlist, rank tracks
 * by how well they'd serve the runner *right now* and build an adaptive
 * "Up Next" queue.
 */

export interface SequencerContext {
  targetSpm: number;
  /** Track currently playing (excluded from ranking, used for transition cost). */
  currentTrack?: Track | null;
  /** Track ids played recently, most recent last. */
  recentlyPlayedIds?: string[];
  /** Track ids the user skipped this session. */
  skippedIds?: string[];
  bpmPrefs?: BpmMatchPreferences;
  weights?: Partial<SequencerWeights>;
}

export interface SequencerWeights {
  /** Weight of the BPM compatibility score. */
  bpmMatch: number;
  /** Max penalty for a track played recently (scaled by recency). */
  recentlyPlayed: number;
  /** Penalty for a track the user skipped this session. */
  skipped: number;
  /** Max penalty for a jarring perceived-BPM jump from the current track. */
  transition: number;
  /** Perceived-BPM jump (SPM) at which the transition penalty saturates. */
  transitionSaturationSpm: number;
  /** Flat score for tracks with unknown BPM (kept, but ranked low). */
  unknownBpmScore: number;
}

export const DEFAULT_SEQUENCER_WEIGHTS: SequencerWeights = {
  bpmMatch: 1,
  recentlyPlayed: 0.45,
  skipped: 0.75,
  transition: 0.2,
  transitionSaturationSpm: 30,
  unknownBpmScore: 0.15,
};

export interface RankedTrack {
  track: Track;
  score: number;
  match: BpmMatch | null;
  reasons: string[];
}

/** Rank every candidate track for the given context, best first. */
export function rankTracks(tracks: Track[], context: SequencerContext): RankedTrack[] {
  const weights = { ...DEFAULT_SEQUENCER_WEIGHTS, ...context.weights };
  const recent = context.recentlyPlayedIds ?? [];
  const skipped = new Set(context.skippedIds ?? []);
  const currentMatch = context.currentTrack
    ? getBpmMatch(context.targetSpm, context.currentTrack.bpm, context.bpmPrefs)
    : null;

  const ranked: RankedTrack[] = [];
  for (const track of tracks) {
    if (context.currentTrack && track.id === context.currentTrack.id) continue;

    const reasons: string[] = [];
    const match = getBpmMatch(context.targetSpm, track.bpm, context.bpmPrefs);
    let score: number;

    if (match === null) {
      score = weights.unknownBpmScore;
      reasons.push("BPM unknown");
    } else {
      score = match.score * weights.bpmMatch;
      reasons.push(
        `${match.perceivedBpm.toFixed(0)} perceived BPM (${match.multiplier}x) vs ${context.targetSpm} SPM target`,
      );

      // Avoid whiplash transitions between consecutive tracks.
      if (currentMatch) {
        const jump = Math.abs(match.perceivedBpm - currentMatch.perceivedBpm);
        const transitionPenalty =
          Math.min(1, jump / weights.transitionSaturationSpm) * weights.transition;
        if (transitionPenalty > 0.01) {
          score -= transitionPenalty;
          reasons.push(`transition jump ${jump.toFixed(0)} BPM from current track`);
        }
      }
    }

    const recentIndex = recent.lastIndexOf(track.id);
    if (recentIndex !== -1) {
      // Most recently played ⇒ full penalty; decays for older plays.
      const recency = (recentIndex + 1) / recent.length;
      const penalty = weights.recentlyPlayed * recency;
      score -= penalty;
      reasons.push("played recently");
    }

    if (skipped.has(track.id)) {
      score -= weights.skipped;
      reasons.push("skipped this session");
    }

    ranked.push({ track, score, match, reasons });
  }

  ranked.sort((a, b) => b.score - a.score);
  return ranked;
}

/** Build the visible "Up Next" queue (top N ranked tracks). */
export function buildQueue(tracks: Track[], context: SequencerContext, length = 5): RankedTrack[] {
  return rankTracks(tracks, context).slice(0, length);
}

/** Pick the next track to play (or null if nothing is available). */
export function pickNextTrack(tracks: Track[], context: SequencerContext): RankedTrack | null {
  return rankTracks(tracks, context)[0] ?? null;
}
